// hyper-chess-board/dist/engine/abstract/bot-process.js
// start, stop, and write should be overridden in the derived classes.
export class BotProcess {
    // log is all of the input/output to/from the engine so far
    // input is indexed by a " > " before the line.
    log = "";
    readListeners = [];
    // for prompt
    promptPrefix;
    onPromptSuccess;
    promptTimeout;
    constructor() { }
    #onReadLine(line) {
        for (const l of this.readListeners)
            l(line);
    }
    addReadLineListener(listener) {
        this.readListeners.push(listener);
    }
    write(cmd) {
        this.log += `> ${cmd}\n`;
    }
    // have you tried turning it off and on again?
    restart() {
        this.stop();
        this.start();
    }
    readLine(line) {
        this.log += `${line}\n`;
        this.#onReadLine(line);
        if (this.promptPrefix && line.startsWith(this.promptPrefix)) {
            if (this.onPromptSuccess)
                this.onPromptSuccess(line);
            clearTimeout(this.promptTimeout);
            delete this.promptPrefix;
        }
    }
    prompt(cmd, prefix, timeoutMs = undefined) {
        if (this.promptPrefix)
            throw new Error("AbstractBotProcess: cannot prompt; currently responding to an earlier prompt");
        return new Promise((res, rej) => {
            this.promptPrefix = prefix;
            this.onPromptSuccess = (line) => res(line);
            if (timeoutMs !== undefined) {
                this.promptTimeout = setTimeout(() => {
                    console.error(`Prompt ${cmd} failed to achieve prefix ${prefix} after ${timeoutMs}ms`);
                    console.error(this.log);
                    delete this.promptPrefix;
                    rej();
                }, timeoutMs);
            }
            this.write(cmd);
        });
    }
}
