// hyper-chess-board/dist/engine/abstract/protocol.js
export class BotProtocol {
    bot;
    thinkStats = {};
    currDepth = -1;
    thinkStatsUpdateListeners = [];
    engineName = "???";
    authorName = "???";
    constructor(botProcess) {
        this.bot = botProcess;
        this.bot.start();
    }
    setEngineName(name) {
        this.engineName = name;
    }
    setAuthorName(name) {
        this.authorName = name;
    }
    getEngineName() {
        return this.engineName;
    }
    getAuthorName() {
        return this.authorName;
    }
    addThinkStatsUpdateListener(listener) {
        this.thinkStatsUpdateListeners.push(listener);
    }
    updateThinkStats(stats) {
        if (stats.depth)
            this.currDepth = stats.depth;
        const currStats = this.thinkStats[this.currDepth];
        if (currStats) {
            Object.assign(currStats, stats);
        }
        else {
            this.thinkStats[this.currDepth] = stats;
        }
        for (const l of this.thinkStatsUpdateListeners)
            l(this.thinkStats[this.currDepth]);
    }
    getThinkStats() {
        return this.thinkStats[this.currDepth] || {};
    }
    resetThinkStats() {
        this.thinkStats = {};
    }
}
