// hyper-chess-board/dist/game/piece.js
// Contains helper functions for pieces, such as their numerical representations
// FEN to piece conversions are also stored here.
export var Side;
(function (Side) {
    Side[Side["None"] = 0] = "None";
    Side[Side["White"] = 8] = "White";
    Side[Side["Black"] = 16] = "Black";
})(Side || (Side = {}));
;
export var PieceType;
(function (PieceType) {
    PieceType[PieceType["None"] = 0] = "None";
    PieceType[PieceType["King"] = 1] = "King";
    PieceType[PieceType["Retractor"] = 2] = "Retractor";
    PieceType[PieceType["Chameleon"] = 3] = "Chameleon";
    PieceType[PieceType["Springer"] = 4] = "Springer";
    PieceType[PieceType["Coordinator"] = 5] = "Coordinator";
    PieceType[PieceType["Straddler"] = 6] = "Straddler";
    PieceType[PieceType["Immobilizer"] = 7] = "Immobilizer";
})(PieceType || (PieceType = {}));
;
const pieceSideMask = 0b11000;
const pieceTypeMask = 0b00111;
export function getPieceType(p) {
    return p & pieceTypeMask;
}
export function getPieceSide(p) {
    return p & pieceSideMask;
}
export function setPieceType(p, t) {
    return getPieceSide(p) | t;
}
export function isPieceOfType(p, o) {
    return getPieceType(p) == o;
}
export function isPieceOfSide(p, o) {
    return getPieceSide(p) == o;
}
export function arePiecesSameSide(p1, p2) {
    return getPieceSide(p1) == getPieceSide(p2);
}
export function arePiecesSameType(p1, p2) {
    return getPieceType(p1) == getPieceType(p2);
}
export const PieceStrings = ["?", "K", "Q", "B", "N", "R", "P", "U"];
const FENCharToPiece = {
    k: Side.Black | PieceType.King,
    q: Side.Black | PieceType.Retractor,
    b: Side.Black | PieceType.Chameleon,
    n: Side.Black | PieceType.Springer,
    r: Side.Black | PieceType.Coordinator,
    p: Side.Black | PieceType.Straddler,
    u: Side.Black | PieceType.Immobilizer,
    K: Side.White | PieceType.King,
    Q: Side.White | PieceType.Retractor,
    B: Side.White | PieceType.Chameleon,
    N: Side.White | PieceType.Springer,
    R: Side.White | PieceType.Coordinator,
    P: Side.White | PieceType.Straddler,
    U: Side.White | PieceType.Immobilizer
};
const PieceTypeToFENChar = {
    [PieceType.King]: "k",
    [PieceType.Retractor]: "q",
    [PieceType.Chameleon]: "b",
    [PieceType.Springer]: "n",
    [PieceType.Coordinator]: "r",
    [PieceType.Straddler]: "p",
    [PieceType.Immobilizer]: "u"
};
export function getFENCharFromPieceType(p) {
    return PieceTypeToFENChar[p];
}
export function getPieceFromFENChar(char) {
    return FENCharToPiece[char];
}
export function getSANCharFromPieceType(p) {
    return PieceStrings[p];
}
