// hyper-chess-board/dist/graphics/widgets/engine-widget.js
import { BoardWidget } from "./board-widget.js";
import { Board, Side } from "../../index.js";
import { WebBotProcess } from "../../engine/web/web-bot-process.js";
import { UCIBotProtocol } from "../../engine/protocols/uci-protocol.js";
export class EngineWidget extends BoardWidget {
    engine = new WebBotProcess("./scripts/hyper-active/main.js");
    protocol = new UCIBotProtocol(this.engine);
    container;
    activeElem;
    evalElem;
    pvElem;
    depthElem;
    npsElem;
    constructor(boardgfx, location) {
        super(boardgfx);
        const container = document.createElement("div");
        container.classList.add("board-graphics__engine");
        container.innerHTML = `
            <input class = "engine__active" type = "checkbox">
            <span class = "engine__eval">-</span>
            <span class = "engine__name">Hyper Chess Engine</span>
            depth
            <span class = "engine__depth"></span>
            (<span class = "engine__nps"></span>nps)
            <div class = "engine__pv">-</div>`;
        boardgfx.getWidgetContainer(location).appendChild(container);
        this.container = container;
        this.setName();
        this.activeElem = this.container.getElementsByClassName("engine__active")[0];
        this.evalElem = this.container.getElementsByClassName("engine__eval")[0];
        this.pvElem = this.container.getElementsByClassName("engine__pv")[0];
        this.depthElem = this.container.getElementsByClassName("engine__depth")[0];
        this.npsElem = this.container.getElementsByClassName("engine__nps")[0];
        this.activeElem.addEventListener("change", () => {
            if (this.activeElem.checked)
                this.start();
            else
                this.stop();
        });
        boardgfx.skeleton.addEventListener("variation-change", (event) => {
            this.startThinking();
        });
    }
    enable() {
        this.container.style.display = "";
    }
    disable() {
        this.container.style.display = "none";
        this.stop();
    }
    start() {
        this.activeElem.checked = true;
        this.setName();
        this.engine.start();
        this.protocol.addThinkStatsUpdateListener((stats) => {
            // get eval
            if (stats.score) {
                const isWTP = this.boardgfx.getTurn() == Side.White;
                const score = (isWTP ? 1 : -1) * stats.score.value;
                const sign = score > 0 ? "+" : "";
                if (stats.score.isMate) {
                    this.evalElem.innerText = `#${sign}${(score < 0 ? -1 : 1) * Math.floor((Math.abs(score) + 1) / 2)}`;
                }
                else {
                    this.evalElem.innerText = `${sign}${(score / 100).toFixed(1)}`;
                }
            }
            // get pv
            if (stats.pv) {
                const pv = stats.pv.split(" ");
                const b = new Board(this.boardgfx.getFEN());
                let fullmove = this.boardgfx.getFullMove();
                let pvSan = this.boardgfx.getTurn() == Side.White ? "" : `${fullmove}... `;
                for (const m of pv) {
                    const move = b.getMoveOfLAN(m);
                    if (b.getTurn() == Side.White)
                        pvSan += `${fullmove}. `;
                    else
                        fullmove++;
                    pvSan += b.getMoveSAN(move) + " ";
                    b.makeMove(move);
                }
                this.pvElem.innerText = pvSan;
            }
            // get depth
            if (stats.depth)
                this.depthElem.innerText = stats.depth.toString();
            // get nps
            if (stats.nodes && stats.time) {
                const nps = stats.nodes / (stats.time / 1000);
                // prefixes determined by every one thousand
                const prefixes = ["", "K", "M", "G"];
                const prefixIdx = Math.floor(Math.log(nps) / Math.log(1000));
                const nptime = (nps / Math.pow(1000, prefixIdx)).toPrecision(3);
                this.npsElem.innerText = `${nptime} ${prefixes[prefixIdx]}`;
            }
        });
        this.startThinking();
    }
    stop() {
        this.activeElem.checked = false;
        if (this.engine.isRunning)
            this.engine.stop();
    }
    startThinking() {
        if (this.engine.isRunning) {
            this.protocol.stopThink();
            this.protocol.setFEN(this.boardgfx.getFEN());
            this.protocol.startThink();
        }
    }
    async setName() {
        const nameElem = this.container.getElementsByClassName("engine__name")[0];
        if (await this.protocol.isReady(10000))
            nameElem.innerText = this.protocol.getEngineName();
    }
}
