// hyper-chess-board/dist/engine/web/web-bot-process.js
import { BotProcess } from "../abstract/bot-process.js";
// Handles writing to or reading from a WASM instance of the engine.
export class WebBotProcess extends BotProcess {
    workerPath;
    worker;
    constructor(workerPath) {
        super();
        this.workerPath = workerPath;
        this.start();
    }
    get isRunning() {
        return this.worker !== undefined;
    }
    start() {
        if (this.worker)
            this.stop();
        this.worker = new Worker(this.workerPath);
        this.worker.onerror = (err) => {
            this.stop();
            throw new Error(`Could not start hyper chess bot web worker: ${err.message}`);
        };
        this.worker.onmessageerror = (err) => {
            throw err;
        };
        this.worker.onmessage = (event) => {
            this.readLine(event.data);
        };
    }
    stop() {
        if (!this.worker)
            return;
        this.worker.terminate();
        delete this.worker;
    }
    write(cmd) {
        if (!this.isRunning)
            return;
        super.write(cmd);
        this.worker.postMessage(cmd);
    }
}
